### PR TITLE
#654 Always call afterNotify

### PR DESCRIPTION
**Pull request by subzero10**
## Status
**READY**

## Description
Issue #654.
Make sure to always call `beforeNotify` and `afterNotify` handlers, even if report is not sent to Honeybadger.

## Todos
- [x] Always call `afterNotify`
- [x] Always call `beforeNotify`
- [x] Lambda functions: should the callback always be called in `afterNotify`? (as it was before #650)
- [x] Tests (`afterNotify` in notice, multiple `afterNotify`, `afterNotify` in notice and in client instance, `afterNotify` in lambda functions)
- [x] Changelog Entry (unreleased)

## Steps to Test or Reproduce
1. Create an HB instance without setting `apiKey`.
2. Set an `afterNotify` handler: client.afterNotify((err) => console.log(err.message))
3. Call client.notify('should not send report')
4. The `afterNotify` handler should be called with an error saying that `apiKey` was not set.

Front logo [Front conversations](https://app.frontapp.com/open/top_3u7vl)

### CODE REVIEW

File: src/core/util.ts [hunk 1]
```diff
@@ -93,6 +93,10 @@ export function runBeforeNotifyHandlers(notice, handlers: BeforeNotifyHandler[])
 }
 
 export function runAfterNotifyHandlers(notice, handlers: AfterNotifyHandler[], error = undefined): boolean {
+  if (notice && notice.afterNotify) {
+    handlers.unshift(notice.afterNotify)
```
Comment (joshuap) [inline src/core/util.ts:97]: Do you want to duplicate `handlers` before mutating it? I believe that's why we were doing the `Array.prototype.slice.call` dance. I don't know if it's necessary, though.

**Comment by joshuap** [comment]
Will `beforeNotify` handlers also always be called?

File: src/core/util.ts (hunk 1 above)
Comment (subzero10) [reply src/core/util.ts:97]: I don't see why they were being duplicated though 🤷 .
One theory is that the array could be mutated from the initial call to `notify` until the `runAfterNotifyHandlers` is called, but that scenario is unlikely to happen.
So, I decided to use them in the same way we use the `runBeforeNotifyHandlers()`.
Should I clone them just to be on the safe side?

**Comment by subzero10** [comment]
> Will `beforeNotify` handlers also always be called?

Good question. At the moment, they are not always called.
For example, if a pre-condition fails (i.e. api key not set), they are not called.
`beforeNotify` handlers are used to add/modify data in `notice` and/or decide whether the `notice` will be sent to Honeybadger, right? If the `notice` is not going to be sent to Honeybadger, is there any point at all to call them?

On the other hand, you could argue that we are calling `afterNotify` without actually _notifying_ Honeybadger, so we should do the same for `beforeNotify`, regardless whether we intend to notify Honeybadger or not.

What do you think?

**Comment by subzero10** [comment]
@joshuap This is ready, unless you want to discuss and modify based on the above.

File: src/core/util.ts (hunk 1 above)
Comment (joshuap) [reply src/core/util.ts:97]: Nope, I think it's fine as-is! I don't remember why they were duplicated either.

**Comment by joshuap** [comment]
> > Will `beforeNotify` handlers also always be called?
> 
> Good question. At the moment, they are not always called. For example, if a pre-condition fails (i.e. api key not set), they are not called. `beforeNotify` handlers are used to add/modify data in `notice` and/or decide whether the `notice` will be sent to Honeybadger, right? If the `notice` is not going to be sent to Honeybadger, is there any point at all to call them (and waste resources)?
> 
> On the other hand, you could argue that we are calling `afterNotify` without actually _notifying_ Honeybadger, so we should do the same for `beforeNotify`, regardless whether we intend to notify Honeybadger or not.
> 
> What do you think?

I think we should run both sets of hooks always. That's what we do in Ruby—although, we also [stop running the `before_notify` hooks if one returns `false`/ignores the notice](https://github.com/honeybadger-io/honeybadger-ruby/blob/5650ecbd5d5a0e8db9324aa01f9fda3ee1e20cc3/lib/honeybadger/agent.rb#L147). I think it probably makes more sense to run all of the hooks in JavaScript and skip the notice if one or more returned false.

**Review by joshuap — COMMENTED** [review]
This is looking good. Let's make the `beforeNotify` handlers change (if you think it makes sense) in this PR, since all of this logic sort of goes together.

File: src/server/middleware.ts [hunk 2]
```diff
@@ -45,28 +45,37 @@ type LambdaHandler = (event: unknown, context: unknown, callback: unknown) => vo
...
+     * we don't have to do that if the handler has third callback function parameter,
+     * because it will be called directly from inside the handler.
+     */
     const shouldInvokeCallbackExplicitly = handler.length < 3
 
+    /**
+     * this flag prevents an infinite loop in case of an error thrown
+     * inside the domain error handler {@link hbHandler}
+     */
+    let domainErrorHandlerAlreadyCalled = false;
```
Comment (subzero10) [inline src/server/middleware.ts:60]: @joshuap Just making sure you saw this.

**Comment by subzero10** [comment]
> This is looking good. Let's make the `beforeNotify` handlers change (if you think it makes sense) in this PR, since all of this logic sort of goes together.

Done. It makes sense to apply this in Javascript if the behavior already exists in Ruby.

Note: There is a potential issue with current integrations. From now on, `beforeNotify` handlers can be called without a notice object (in case someone calls `honeybadger.notify({})`). I documented this change in Typescript by modifying the `notice` in `BeforeNotifyHandler` to be optional:
```
export interface BeforeNotifyHandler {
    (notice?: Notice): boolean | void;
}
```

**Review by joshuap — APPROVED** [review]
Looks good to me. This may be a major release due to the changed behavior—I'll have to think about that. I may also release a beta version first so that we can try it out first.

**Comment by subzero10** [comment]
> Looks good to me. This may be a major release due to the changed behavior—I'll have to think about that. I may also release a beta version first so that we can try it out first.

Any updates on this? Have you decided whether it is going to be a major release?

**Comment by joshuap** [comment]
@subzero10 yeah, I'm leaning towards a minor release (3.3), after I finish up #670. I'll probably release a beta in December and do the actual release in January.

**Pull request merged**